mips/rootdev.c

```diff
@@ -22,28 +28,97 @@ static void rootdev_intr_teardown(device_t *dev, intr_handler_t *handler) {
 }
 
-extern pci_bus_driver_t gt_pci_bus;
-device_t *gt_pci;
-
 static int rootdev_attach(device_t *dev) {
+  rman_create(&rm_mem, MALTA_PHYS_ADDR_SPACE_BASE, MALTA_PHYS_ADDR_SPACE_END);
+
   gt_pci = device_add_child(dev);
   gt_pci->driver = &gt_pci_bus.driver;
   if (device_probe(gt_pci))
     device_attach(gt_pci);
   return 0;
 }
 
+static uint8_t mips_read_1(resource_t *handle, unsigned offset) {
+  intptr_t addr = handle->r_start + offset;
+  return *(volatile uint8_t *)MIPS_PHYS_TO_KSEG1(addr);
+}
+
+static void mips_write_1(resource_t *handle, unsigned offset, uint8_t value) {
+  intptr_t addr = handle->r_start + offset;
+  *(volatile uint8_t *)MIPS_PHYS_TO_KSEG1(addr) = value;
+}
+
+static uint16_t mips_read_2(resource_t *handle, unsigned offset) {
+  intptr_t addr = handle->r_start + offset;
+  return *(volatile uint16_t *)MIPS_PHYS_TO_KSEG1(addr);
+}
+
+static void mips_write_2(resource_t *handle, unsigned offset, uint16_t value) {
+  intptr_t addr = handle->r_start + offset;
+  *(volatile uint16_t *)MIPS_PHYS_TO_KSEG1(addr) = value;
+}
+
+static uint32_t mips_read_4(resource_t *handle, unsigned offset) {
+  intptr_t addr = handle->r_start + offset;
+  return *(volatile uint32_t *)MIPS_PHYS_TO_KSEG1(addr);
+}
+
+static void mips_write_4(resource_t *handle, unsigned offset, uint32_t value) {
+  intptr_t addr = handle->r_start + offset;
+  *(volatile uint32_t *)MIPS_PHYS_TO_KSEG1(addr) = value;
+}
+
+static void mips_read_region_1(resource_t *handle, unsigned offset,
+                               uint8_t *dst, size_t count) {
+  uint8_t *src = (uint8_t *)MIPS_PHYS_TO_KSEG1(handle->r_start + offset);
+  for (size_t i = 0; i < count; i++)
+    *dst++ = *src++;
+}
+
+static void mips_write_region_1(resource_t *handle, unsigned offset,
+                                const uint8_t *src, size_t count) {
+  uint8_t *dst = (uint8_t *)MIPS_PHYS_TO_KSEG1(handle->r_start + offset);
+  for (size_t i = 0; i < count; i++)
+    *dst++ = *src++;
+}
+
+static struct bus_space generic_space = {.read_1 = mips_read_1,
+                                         .write_1 = mips_write_1,
+                                         .read_2 = mips_read_2,
+                                         .write_2 = mips_write_2,
+                                         .read_4 = mips_read_4,
+                                         .write_4 = mips_write_4,
+                                         .read_region_1 = mips_read_region_1,
+                                         .write_region_1 = mips_write_region_1};
+
+static resource_t *rootdev_resource_alloc(device_t *bus, device_t *child,
+                                          int type, int rid, rman_addr_t start,
+                                          rman_addr_t end, size_t size,
+                                          unsigned flags) {
+
+  resource_t *r =
+    rman_allocate_resource(&rm_mem, start, end, size, size, RF_NONE);
+  if (r == NULL)
+    return NULL;
+
+  r->r_owner = child;
```

Comment:
That is not very different from some lines in `gt_pci_resource_alloc`. Shouldn't that be extracted to `bus_generic_resource_alloc`?

Reply:
Did  you mean `bus_resource_alloc` by `bus_generic_resource_alloc`, or some new function?

Reply:
A new function that contains a common piece of code used by various implementations of `bus_resource_alloc` in bus drivers. Refer to [bus_generic_probe](https://github.com/cahirwpz/mimiker/blob/master/sys/device.c#L49).